### PR TITLE
Fix to #18492 - Query: inject SearchConditionConvertingExpressionVisitor into ParameterValueBasedSelectExpressionOptimizer

### LINKED ISSUE #18492: Query: inject SearchConditionConvertingExpressionVisitor into ParameterValueBasedSelectExpressionOptimizer
When performing optimizations in ParameterValueBasedSelectExpressionOptimizer we need to run search condition converter for SQL Server, since some of the optimizations can reduce predicate to a simple bool constant (incorrect on SQL Server). 

Currently we have a workaround that performs optimizations and if the expression is a SelectExpression, peeks into its predicate to see if it was changed to SqlConstantExpression. If so, convert it to `1 = 1` or `0 = 1`. This is only necessary on SQL Server but currently we do it for all providers. Once ParameterValueBasedSelectExpressionOptimizer is provided by DI we should only do it where needed.

### LINKED ISSUE #18940: Query: copy-paste error in the SelectExpression Equals method, could lead to null refs/other errors
SelectExpression -> private bool Equals(SelectExpression selectExpression)

code:

```cs
            if (!(Having == null && selectExpression.Having == null
                || Having != null && Predicate.Equals(selectExpression.Having)))
            {
                return false;
            }
```

should be:

```cs
            if (!(Having == null && selectExpression.Having == null
                || Having != null && Having.Equals(selectExpression.Having)))
            {
                return false;
            }
```

### PR DESCRIPTION
**Pull request by maumar**
Using DI to create parameter value based postprocessor, which allows for provider-specific optimizations, including SearchConditionConvertingExpressionVisitor

Resolves #18492
Resolved #18940

### CODE REVIEW

File: src/EFCore.Relational/Query/IRelationalParameterBasedQueryPostprocessorFactory.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.EntityFrameworkCore.Query.Internal;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.EntityFrameworkCore.Query
+{
+    /// <summary>
+    ///     <para>
+    ///         A factory for creating <see cref="RelationalParameterBasedQueryPostprocessor" /> instances.
```
Comment (maumar) [inline src/EFCore.Relational/Query/IRelationalParameterBasedQueryPostprocessorFactory.cs]: there is some inconsistency here - sometimes we point to the factory and sometimes to the created types - which one should it be?

File: src/EFCore.Relational/Query/Internal/RelationalParameterBasedQueryPostprocessorDependencies.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Storage;
+using Microsoft.EntityFrameworkCore.Utilities;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.EntityFrameworkCore.Query.Internal
+{
+    /// <summary>
+    ///     <para>
+    ///         Service dependencies parameter class for <see cref="RelationalParameterBasedQueryPostprocessorFactory" />
```
Comment (maumar) [inline src/EFCore.Relational/Query/Internal/RelationalParameterBasedQueryPostprocessorDependencies.cs]: there is some inconsistency here - sometimes we point to the factory and sometimes to the created types - which one should it be?

File: src/EFCore.Relational/Infrastructure/EntityFrameworkRelationalServicesBuilder.cs [hunk 3]
```diff
@@ -89,7 +89,7 @@ public class EntityFrameworkRelationalServicesBuilder : EntityFrameworkServicesB
                 { typeof(IRelationalTypeMappingSourcePlugin), new ServiceCharacteristics(ServiceLifetime.Singleton, multipleRegistrations: true) },
                 { typeof(IMethodCallTranslatorPlugin), new ServiceCharacteristics(ServiceLifetime.Singleton, multipleRegistrations: true) },
                 { typeof(IMemberTranslatorPlugin), new ServiceCharacteristics(ServiceLifetime.Singleton, multipleRegistrations: true) },
-
+                { typeof(IRelationalParameterBasedQueryPostprocessorFactory), new ServiceCharacteristics(ServiceLifetime.Singleton) }
```
Comment (smitpatel) [inline src/EFCore.Relational/Infrastructure/EntityFrameworkRelationalServicesBuilder.cs]: QueryTranslationPostprocessor?

File: src/EFCore.SqlServer/Query/Internal/SqlServerParameterBasedQueryPostprocessor.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            bool useRelationalNulls)
+            : base(dependencies, useRelationalNulls)
+        {
+        }
+
+        public override (SelectExpression selectExpression, bool canCache) Optimize(
+            SelectExpression selectExpression,
+            IReadOnlyDictionary<string, object> parametersValues)
+        {
+            var result = base.Optimize(selectExpression, parametersValues);
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerParameterBasedQueryPostprocessor.cs]: tuple deconstruction.

File: src/EFCore.Relational/Query/Internal/RelationalParameterBasedQueryPostprocessorDependencies.cs (hunk 2 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/Internal/RelationalParameterBasedQueryPostprocessorDependencies.cs]: @AndriySvyryd @ajcvickers

File: src/EFCore.Relational/Query/Internal/RelationalParameterBasedQueryTranslationPostprocessor.cs [hunk 5]
```diff
@@ -20,41 +19,39 @@ namespace Microsoft.EntityFrameworkCore.Query.Internal
     ///     any release. You should only use it directly in your code with extreme caution and knowing that
     ///     doing so can result in application failures when updating to a new Entity Framework Core release.
     /// </summary>
-    public class ParameterValueBasedSelectExpressionOptimizer
+    public class RelationalParameterBasedQueryTranslationPostprocessor
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/RelationalParameterBasedQueryTranslationPostprocessor.cs]: Move out of internal

**Pull request merged**